File: jump-i.java
// Memoization

// Time - O(n^2)
// Space - O(n) // Stack

// class Solution {
//     public boolean canJump(int[] nums) {
//         int[] dp = new int[nums.length];
//         return helper(0, nums, dp);
//     }
    
//     private boolean helper(int i, int[] nums, int[] dp){
        
//         if(i >= nums.length -1){
//             return true;
//         }
        
//         if(dp[i] != 0){
//             return dp[i] == 1; 
//         }
        
//         for(int j = 1; j <= nums[i]; j++ ){
//             boolean result = helper(i+j, nums, dp);
//             if(result == true){
//                 dp[i] = 1;
//                 return true;
//             }
//         }
        
//         dp[i] = -1;
//         return false;
//     }
// }


// Greedy
//Time - O(n^2)
//Space - O(n)


class Solution {
    public boolean canJump(int[] nums) {
        int destination = nums.length - 1;
        
        for(int i = nums.length - 2; i >= 0; i--){
            if(nums[i] >= destination - i){
                destination = i;
            }
        }
        return destination == 0;
    }
}

File: jumpI.py
'''
DFS approach can be applied here
Time Complexity --> O(n) since it can travel to all noeds
Space Complexity --> O(n) that is the stack space it will occupy
'''
class Solution:
    def canJump(self, nums: List[int]) -> bool:

        def dfs(index, nums):
            if index == len(nums)-1:
                return True
            #logic
            visited[index] = 1
            #print(visited)
            for i in range(1, nums[index]+1):
                new_ind = index+i
                if new_ind not in visited and dfs(new_ind, nums):
                    return True
            return False

                        
                        
                
            
        visited = {}
        return dfs(0, nums)

        



